File: MagicIndex.java
import java.util.*;
import java.io.*;

public class MagicIndex
{
    private static ArrayList<Integer> list = new ArrayList<Integer>();

    public static int findMagicIndex(int low, int high)
    {
        if (low <= high) {
            int mid = (low + high)/2;
            if (list.get(mid) > mid) {
                return findMagicIndex(low, mid-1);
            } else if (list.get(mid) < mid) {
                return findMagicIndex(mid +1 , high);
            } else if (list.get(mid) == mid){
                return mid;
            }
        }
        return -1;
    }

    public static void main(String[] args)
    {
        Scanner sc = new Scanner(System.in);

        while(sc.hasNextInt()) {
            list.add(sc.nextInt());
        }

        System.out.println("Magic Index is : " + findMagicIndex(0, list.size()-1));

        sc.close();
    }
}

File: C/magicIndex.c
#include<stdio.h>

int main()
{
   int a[]={0,-2,-1,2,4,7,8};
   int len=sizeof(a)/sizeof(int);
   int i;

   int low=1; int high=len-1; int mid=(low+high)/2;

   while(low<=high){
        if(a[mid]==mid){
            printf("%d\n",mid);
            break;
        }
        else if(a[mid] >mid){
            high=mid-1;
            mid=(low+high)/2;
        }
        else {
            low=mid+1;
            mid=(low+high)/2;
        }
   }

   return 0;
}
